src/ByteCountingStream.php

```diff
@@ -0,0 +1,58 @@
+<?php
+namespace GuzzleHttp\Psr7;
+
+use Psr\Http\Message\StreamInterface;
+
+/**
+ *  Stream decorator that counts the number of bytes read off of an underlying
+ *  read stream.
+ */
+class ByteCountingStream implements StreamInterface
+{
+    use StreamDecoratorTrait;
+
+    /** @var int Number of bytes allowed per read */
+    private $byteCount;
+
+    /**
+     * @param StreamInterface $stream Stream to wrap
+     * @param int             $bytes  Number of bytes per read
+     * @throws \InvalidArgumentException
+     */
+    public function __construct(StreamInterface $stream, $bytes)
+    {
+        $this->stream = $stream;
+
+        if (!is_int($bytes) || $bytes < 0) {
+            $msg = "Bytes per read should be non-negative integer, got {$bytes}.";
```

Comment:
Maybe be careful if bytes is non-scalar when trying to put it in here. Say if someone passed through a closure, you'd get a fatal error.